File: Data Structures & Algorithms/design-hit-counter/submission-1.cpp
class HitCounter {
private:
    queue<int> hits;
public:
    HitCounter() {
        
    }
    
    void hit(int timestamp) {
        hits.push(timestamp);
    }
    
    int getHits(int timestamp) {
        while (!hits.empty()) {
            int diff = timestamp - hits.front();
            if (diff >= 300) hits.pop();
            else break;
        }
        return hits.size();
    }
};

/*
hitCounter.hit(1);       // hit at timestamp 1.
hitCounter.hit(2);       // hit at timestamp 2.
hitCounter.hit(3);       // hit at timestamp 3.
hitCounter.getHits(4);   // get hits at timestamp 4, return 3.
hitCounter.hit(300);     // hit at timestamp 300.
hitCounter.getHits(300); // get hits at timestamp 300, return 4.
hitCounter.getHits(301); // get hits at timestamp 301, return 3.



t1, t2, t3, t300

getHits(300) -> [0, 300], all need > 0 and <= 300
getHits(301) -> [1, 301], all need > 1 and <= 300



*/

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter* obj = new HitCounter();
 * obj->hit(timestamp);
 * int param_2 = obj->getHits(timestamp);
 */

File: Data Structures & Algorithms/maximum-frequency-stack/submission-1.py
class FreqStack:

    def __init__(self):
        self.count = {}
        self.max_count = 0
        self.stacks = {}

    def push(self, val: int) -> None:
        valcount = 1 + (self.count[val] if val in self.count else 0)
        # valcount = 1 + self.count.get(val, 0)
        self.count[val] = valcount

        if valcount > self.max_count:
            self.max_count = valcount
            self.stacks[valcount] = []
        
        self.stacks[valcount].append(val)

    def pop(self) -> int:
        res = self.stacks[self.max_count].pop()
        self.count[res] -= 1

        if not self.stacks[self.max_count]:
            self.max_count -= 1
        
        return res


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()